### PR TITLE
using NamespacedId codec to serialize the response from metadata search

### PR DESCRIPTION
**Pull request by gsps1**

### CODE REVIEW

**Comment by bdmogal** [comment]
LGTM, but can we add a unit test, so we can catch this even earlier?

**Comment by gsps1** [comment]
sure, have added a test case, and added one minor bugfix for the case when target=ALL in query param for search.

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 1]
```diff
@@ -141,6 +143,28 @@ public void testTags() throws IOException {
     tags = getTags(mystream);
     Assert.assertTrue(tags.containsAll(streamTags));
     Assert.assertTrue(streamTags.containsAll(tags));
+    // test search for stream
+    Set<MetadataSearchResultRecord> searchTags = searchTags(Id.Namespace.DEFAULT.getId(), "stT", "STREAM");
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: is it possible to use ``MetadataSearchType`` here?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 2]
```diff
@@ -141,6 +143,28 @@ public void testTags() throws IOException {
...
+    Set<MetadataSearchResultRecord> expected = ImmutableSet.of(
+      new MetadataSearchResultRecord(mystream, MetadataSearchTargetType.STREAM)
+    );
+    Assert.assertEquals(expected.size(), searchTags.size());
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: can't we just ``Assert.assertEquals(expected, searchTags)``

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 3]
```diff
@@ -141,6 +143,28 @@ public void testTags() throws IOException {
...
+    Assert.assertTrue(searchTags.containsAll(expected));
+
+    // test prefix search, should match stream and service programs
+    searchTags = searchTags(Id.Namespace.DEFAULT.getId(), "s", "ALL");
+    expected = ImmutableSet.of(
+      new MetadataSearchResultRecord(mystream, MetadataSearchTargetType.STREAM),
+      new MetadataSearchResultRecord(pingService, MetadataSearchTargetType.PROGRAM)
+    );
+    Assert.assertEquals(expected.size(), searchTags.size());
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: ditto

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 4]
```diff
@@ -141,6 +143,28 @@ public void testTags() throws IOException {
...
+    Assert.assertTrue(searchTags.containsAll(expected));
+
+    // search without any target param
+    searchTags = searchTags(Id.Namespace.DEFAULT.getId(), "s", null);
+    Assert.assertEquals(expected.size(), searchTags.size());
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: ditto

**Comment by bdmogal** [comment]
Should we add tests for searching properties as well?

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java (hunk 2 above)
Comment (gsps1) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: that should be enough.

**Comment by gsps1** [comment]
@bdmogal have added coverage for properties and addressed your comment.

**Comment by henry-cask** [comment]
LGTM

File: cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java [hunk 5]
```diff
@@ -85,6 +87,26 @@ public void testProperties() throws IOException {
...
+    searchProperties = searchMetadata(Id.Namespace.DEFAULT.getId(), "sKey:s", "ALL");
+    expected = ImmutableSet.of(
+      new MetadataSearchResultRecord(pingService, MetadataSearchTargetType.PROGRAM)
+    );
+    Assert.assertEquals(expected, searchProperties);
+
+    // search without any target param
+    searchProperties = searchMetadata(Id.Namespace.DEFAULT.getId(), "sKey:s", null);
+    Assert.assertEquals(expected, searchProperties);
+
```
Comment (bdmogal) [inline cdap-app-fabric/src/test/java/co/cask/cdap/metadata/MetadataHttpHandlerTest.java]: in both these search verifications, can you add a test that returns empty? a good case would be to search for an existing property/tag, but on the wrong target type.

**Comment by bdmogal** [comment]
@gsps1 one final verification request. Other than that, LGTM.

**Comment by bdmogal** [comment]
Thanks for adding tests!

**Comment by gsps1** [comment]
my build failed due to a flaky test, but https://github.com/caskdata/cdap/pull/4010 - which is based off my branch passed. so merging this.
